### PR TITLE
Add proper support for systemd inside of podman

### PR DESCRIPTION
**Pull request by rhatdan**
Signed-off-by: Daniel J Walsh <dwalsh@redhat.com>

### CODE REVIEW

File: pkg/spec/spec.go [hunk 1]
```diff
@@ -188,6 +188,47 @@ func CreateConfigToOCISpec(config *CreateConfig) (*spec.Spec, error) { //nolint
...
 
+	if strings.HasSuffix(config.Command[0], "init") ||
+		strings.HasSuffix(config.Command[0], "systemd") {
+		options := []string{"rw", "private", "noexec", "nosuid", "nodev"}
+		for _, i := range []string{"/run", "/run/lock", "/sys/fs/cgroup/systemd"} {
+			tmpfsMnt := spec.Mount{
+				Destination: i,
+				Type:        "tmpfs",
+				Source:      "tmpfs",
+				Options:     append(append(options, "tmpcopyup"), "size=65536k"),
```
Comment (mheon) [inline pkg/spec/spec.go]: Don't need a double append, append(options, "tmpcopyup", "size=65536k") should work

File: pkg/spec/spec.go [hunk 2]
```diff
@@ -188,6 +188,47 @@ func CreateConfigToOCISpec(config *CreateConfig) (*spec.Spec, error) { //nolint
...
+		tmpfsMnt := spec.Mount{
+			Destination: spliti[0],
+			Type:        "tmpfs",
+			Source:      "tmpfs",
+			Options:     append(options, "tmpcopyup"),
+		}
+		g.AddMount(tmpfsMnt)
+	}
+
 	for _, i := range config.Tmpfs {
```
Comment (mheon) [inline pkg/spec/spec.go]: Is this still supposed to be here? Looks duplicated

Comment (rhatdan) [reply pkg/spec/spec.go]: Yup, removed.

File: pkg/spec/spec.go (hunk 1 above)
Comment (rhatdan) [reply pkg/spec/spec.go]: Thought that did not work. Fixed.

**Comment by TomSweeneyRedHat** [comment]
bot, retest this please

**Comment by vrothberg** [comment]
bot, retest this please

**Comment by vrothberg** [comment]
Code LGTM but can we document that somewhere in the manpages? Some users might be surprised or may even be looking for exactly this information.

**Comment by vrothberg** [comment]
Btw, I find this to be a very nice feature! I've heard systemd requests quite often and everybody ended up googling what needs to be mounted where.

**Comment by jamescassell** [comment]
I love the feature, but the implementation seems like hard coded magic. Would it be better to have a config option, and/or a configurable regex?

**Comment by rhatdan** [comment]
I suppose to could add the list of directories to libpod.conf, but we really have not been stressing that config file so far.  I just modified the patches to handle the cases where the caller volume mounts in some content into the container at the same location.  So if you volume mount in /run we will use that rather then mount a tmpfs.

I can  add a --systemd flag defaulted to true, so that people could turn off this behavior, for cases where they are running init, but not a systemd init.

**Comment by rhatdan** [comment]
@jamescassell WDYT?
@giuseppe PTAL

**Comment by rhatdan** [comment]
@vrothberg PTAL

**Comment by rhatdan** [comment]
@adelton PTAL

**Comment by rhatdan** [comment]
I would rather leave it default and add configuration to libpod.conf.  I would prefer not to break systemd containers just because a user forgot a flag.

File: cmd/podman/common.go [hunk 3]
```diff
@@ -365,6 +365,10 @@ var createFlags = []cli.Flag{
 		Name:  "sysctl",
 		Usage: "Sysctl options (default [])",
 	},
+	cli.BoolTFlag{
+		Name:  "systemd",
+		Usage: "Run container in systemd mode if the command executable  is systemd or init",
```
Comment (giuseppe) [inline cmd/podman/common.go]: extra space after executable

File: pkg/spec/spec.go [hunk 4]
```diff
@@ -188,6 +189,12 @@ func CreateConfigToOCISpec(config *CreateConfig) (*spec.Spec, error) { //nolint
 		}
 	}
 
+	if config.Systemd && (strings.HasSuffix(config.Command[0], "init") ||
```
Comment (giuseppe) [inline pkg/spec/spec.go]: could we use `path.Base` here instead of `strings.HasSuffix`?

File: pkg/spec/spec.go [hunk 5]
```diff
@@ -328,6 +335,42 @@ func blockAccessToKernelFilesystems(config *CreateConfig, g *generate.Generator)
...
+	options := []string{"rw", "private", "noexec", "nosuid", "nodev"}
+	for _, dest := range []string{"/run", "/run/lock", "/sys/fs/cgroup/systemd"} {
+		if libpod.MountExists(mounts, dest) {
+			continue
+		}
+		tmpfsMnt := spec.Mount{
+			Destination: dest,
+			Type:        "tmpfs",
+			Source:      "tmpfs",
+			Options:     append(options, "tmpcopyup", "size=65536k"),
```
Comment (giuseppe) [inline pkg/spec/spec.go]: tmpcopyup is not standard OCI, it is understood only by runc, could these be just empty tmpfs?

**Comment by jamescassell** [comment]
A config option, with an override via e.g. `--{no-,}systemd` seems sufficient.

File: pkg/spec/spec.go (hunk 5 above)
Comment (rhatdan) [reply pkg/spec/spec.go]: No and we already do this for tmpfs.  I think the oci runtimes should ignore it.
The problem with not copying up is containers that come with a subdir of /run/httpd will blow up if the tmpfs does not have the httpd directory copied up.

File: docs/podman-create.1.md [hunk 6]
```diff
@@ -538,6 +538,17 @@ Network Namespace - current sysctls allowed:
 
 Note: if you use the --network=host option these sysctls will not be allowed.
 
+**--systemd**=*true*|*false*
+
+Run container in systemd mode. The default is *true*.
+
+If the command you running inside of the container is systemd or init, podman
```
Comment (marcov) [inline docs/podman-create.1.md]: maybe you meant **you are running** or **you run**?

File: docs/podman-run.1.md [hunk 7]
```diff
@@ -568,6 +568,17 @@ Network Namespace - current sysctls allowed:
 
 Note: if you use the `--network=host` option these sysctls will not be allowed.
 
+**--systemd**=*true*|*false*
+
+Run container in systemd mode. The default is *true*.
+
+If the command you running inside of the container is systemd or init, podman
```
Comment (marcov) [inline docs/podman-run.1.md]: Same comment as above.

**Comment by adelton** [comment]
If there is something specific needed for systemd to be run in containers, like executing `oci-systemd-hook` to prepare the environment the way systemd expects it, it should really be based on an explicit flag (label?) in the image, not on `podman run` command-line option or by guessing from the entrypoing `basename`.

**Comment by vrothberg** [comment]
@rhatdan Any updates? I just started working with the systemd hook and this PR can make things a bit more convenient for users.

**Comment by rhatdan** [comment]
@vrothberg Sorry I have been on Partial PTO This week, and next.  Just trying to keep up with whats going on but only working a couple of hours each day.

**Comment by vrothberg** [comment]
@rhatdan, I am sorry. No worries, enjoy your PTO! I will be gone the last three weeks of September (but will keep an eye on mails every few days).

**Comment by baude** [comment]
@rhatdan here's a dockerfile to try...
○ → cat Dockerfile 
FROM registry.fedoraproject.org/fedora:28
RUN dnf -y install httpd && dnf clean all && systemctl enable httpd &&  echo "Hello World!" > /var/www/html/index.html
STOPSIGNAL SIGRTMIN+3
EXPOSE 80
CMD ["/usr/sbin/init"]

**Comment by rhatdan** [comment]
Conmon is complaining about Aug 31 12:23:29 localhost.localdomain conmon[27608]: conmon 7cfd36d83da45560ef41 <error>: Failed to get console terminal settings Inappropriate ioctl for device

**Comment by rhatdan** [comment]
Ok this needs the latest version of runc in order to work.  I just released a new version of runc in to F28 updates testing.

**Comment by rhatdan** [comment]
bot, retest this please

**Comment by mheon** [comment]
@rhatdan I think our RPMs require `oci-systemd-hook` now - should we changed that? Also, is there any way to make sure we don't run oci-systemd-hook even if we have it installed for e.g. CRI-O?

**Comment by rhatdan** [comment]
@mheon Lets get this in and then we can remove the requirement.  oci-systemd-hook will just do nothing, if their is a mount point at the location where it wants to mount.  So this should not cause any issues. but we should no longer need to execute the hook.

**Comment by rhatdan** [comment]
We might want to add this feature to cri-o too, once we get this working.

**Comment by baude** [comment]
bot, retest this please

**Comment by mheon** [comment]
LGTM

**Comment by baude** [comment]
LGTM

**Comment by mheon** [comment]
@rh-atomic-bot r+

**Comment by mheon** [comment]
@rh-atomic-bot retry

**Comment by mheon** [comment]
bot, retest this please

**Comment by mheon** [comment]
Alright, that's it. CI is consistently failing even though it says "Exit 0"

I'll look into this more on Tuesday. For now, I'm manually merging.

**Comment by vrothberg** [comment]
> Ok this needs the latest version of runc in order to work. I just released a new version of runc in to F28 updates testing.

There seems to an issue with loading AppArmor profiles (cannot be loaded anymore) and the changes in this PR. I'll take a look.

**Comment by vrothberg** [comment]
> There seems to an issue with loading AppArmor profiles (cannot be loaded anymore) and the changes in this PR. I'll take a look.

Local hiccup. I read some of the comments above that the oci-systemd-hook isn't required anymore (but it is). Can someone shed some light on that?

**Comment by rhatdan** [comment]
oci-systemd-hook should not be required by this patch for podman, but it is still required for CRI-O, we could add similar functionality to CRI-O to eliminate the need.  I am not sure how often CRI-O runs systemd based containers.

**Comment by vrothberg** [comment]
@rhatdan I fear the patch still requires the systemd hook. @giuseppe and I could reproduce both on Fedora and openSUSE.

**Comment by vrothberg** [comment]
@rhatdan Without the `--privileged` we hit the following error:
```
Failed to create /machine.slice/libpod-412f3344c931021326f9901715046552f10430e5580d4d51e56cd5932d2eec63.scope/init.scope control group: No such file or directory
```

**Comment by rhatdan** [comment]
Ok I will look into this, on Tuesday.  (US Holiday today)  I thought I was testing without the hook.

**Comment by giuseppe** [comment]
> Ok I will look into this, on Tuesday. (US Holiday today) I thought I was testing without the hook.

`ctr.CGroupPath()` needs to be bind mounted from the host.  The tmpfs for systemd cgroups is not enough

**Pull request merged**